src/main/java/teammates/storage/search/SearchManager.java

```diff
@@ -37,30 +68,66 @@ public StudentSearchResultBundle searchStudents(String queryString, List<Instruc
         if (!isSearchServiceActive()) {
             throw new SearchNotImplementedException();
         }
-        // TODO
-        throw new SearchNotImplementedException();
+
+        SolrClient client = getSolrClient();
+        QueryResponse response = null;
+        SolrQuery studentQuery = new SolrQuery();
+        studentQuery.setQuery(queryString);
+
+        try {
+            response = client.query(STUDENT_COLLECTION_NAME, studentQuery);
+        } catch (SolrServerException e) {
+            log.severe(String.format(ERROR_SEARCH_DOCUMENT, queryString, e.getRootCause())
+                    + TeammatesException.toStringWithStackTrace(e));
+        } catch (IOException e) {
+            log.severe(String.format(ERROR_SEARCH_DOCUMENT, queryString, e.getCause())
+                    + TeammatesException.toStringWithStackTrace(e));
```

Comment:
Is catch & re-throw acceptable here? (usually an anti-pattern but we want to narrow down the type)

BTW, is it necessary to log the stack-trace?

Reply:
Thinking of implementing a `SearchNotAvailableException` that gives an error message which indicates server errors (e.g. unable to connect to Solr). For re-throw exception I feel that it is okay here.

https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions

Reply:
Also, currently the user will be presented a toast error message stating `Search service is not implemented` if `app.search.service.host` in `build.properties` is not set. But does the user really need to care about whether search is implemented or not? I think they only need to care about whether search is available or not. If search is unavailable, it could mean that:
- search is not implemented
- server could not connect to search engine
- other backend failures (which the user don't really need to care about)

This way we can group all errors propagated to the user into 2 categories:
- Search not available (which includes all the server errors listed above)
- Search result not found

Reply:
To be very technical here, all the cases above correspond to different `5XX` codes:
- Search service is never implemented: `501`; no use retrying
- Search service is available but cannot be connected to: `502`; could possibly retry if the error source is known/fixed
- Search service returns error: also `502`
- Search service is usually available but currently taken down: `503`; no use retrying
- Search service timed out: `504`; could retry
- Search result not found: this is not an error. Just return `200` with an empty list.

Notice that `500` is not considered. `500` tells that the fault is ours, but is not relevant now because the error source is always the external service.

Of course, since we are the sole consumer of the API endpoints we create, it is up to us to interpret those response codes. You're also right to mention that users don't care about technical details; what matters is we give accurate error message. However, these response codes are generally well-defined in the realms of API design and we have no reason to deviate from them.